torch/ao/quantization/dynamic_tracing/auto_trace.py

```diff
@@ -0,0 +1,554 @@
+    # id=1, we know that the output of op_n is the input to op_n+2.
+    qtensor_id = [0]
+    module_id_to_fqn: Dict[int, str] = {}
+
+    class QuantizationInterceptionProxy(torch.Tensor):
+        """
+        An override of `torch.Tensor` to enable dynamic tracing for
+        quantization.
+
+        For each function with a `__torch_function__` override, this proxy does
+        the following for functions which need quantization:
+
+        1. calls `_auto_quant_state.validate_cur_op` to validate that
+           the currently seen op is the same as what was recorded during tracing
+        2. calls `_auto_quant_state.op_prepare_before_hook`
+        3. executes the original function
+        4. calls `_auto_quant_state.op_prepare_after_hook`
+        5. calls `_auto_quant_state.mark_cur_op_complete` to increment
+           the current op index in preparation for the next op
+
+        Otherwise, calls the original function.
+        """
+
+        @classmethod
+        def __torch_function__(cls, func, types, args=(), kwargs=None):
+            # to prevent printing things from going into an infinite loop
+            if func == torch.Tensor.__repr__:
+                return super().__torch_function__(func, types, args, kwargs)
+            if enable_logging:
+                logger.debug(f'__torch_function__ {str(func)} len_args {len(args)}')
+
+            nonlocal qtensor_id
+            nonlocal cur_module
+            kwargs = kwargs if kwargs else {}
+            # if we are in a function, the current module is always a parent
+            parent_module = cur_module
+            hook_type = get_torch_function_hook_type(parent_module, func)
+
+            if hook_type is HookType.OP_HOOKS:
+                assert parent_module is not None
+                qstate = parent_module._auto_quant_state
+                fqn = module_id_to_fqn[id(parent_module)] if parent_module else None
+                if not first_call:
+                    qstate.validate_cur_op(func)
+                # run "before" hook
+                args, kwargs = qstate.op_prepare_before_hook(
+                    func, args, kwargs, first_call, qtensor_id, fqn, parent_module)
+                # forward
+                output = super().__torch_function__(func, types, args, kwargs)
+                # run "after" hook
+                output = qstate.op_prepare_after_hook(
+                    func, output, args, first_call, qtensor_id, parent_module)
+                qstate.mark_cur_op_complete(func)
+            else:
+                output = super().__torch_function__(func, types, args, kwargs)
+
+            # TODO: is this right? Don't really understand this
+            if output is NotImplemented:
+                with torch._C.DisableTorchFunction():
+                    output = func(*args, **kwargs).as_subclass(
+                        QuantizationInterceptionProxy)
+                assert output is not NotImplemented
```

Comment:
what is this for? looks like it's some verification code

Reply:
this is left over from @jamesr66a 's original code, saving for a future PR to debug and hopefully remove this.  Some test cases break without this.

Reply:
`__torch_function__`'s multiple-dispatch system will return `NotImplemented` if none of the `Tensor` subclasses (specified by `types`) dispatches into an implementation and returns a value. See https://github.com/pytorch/rfcs/blob/master/RFC-0001-torch-function-for-methods.md#proposal. So, this section is needed to handle that case. What wasn't clear to me when writing the original code is: when would this ever actually happen given that we always have the concrete `Tensor` class in the dispatch chain? I believe the suggestion for disabling `__torch_function__` dispatch was from @ezyang, and maybe he could shed some more light on the behavior here